model-api/src/commonMain/kotlin/org/modelix/model/api/INodeResolutionScope.kt

```diff
@@ -1,41 +1,113 @@
 package org.modelix.model.api
 
-import kotlin.coroutines.CoroutineContext
+import org.modelix.kotlin.utils.ContextValue
 
-interface INodeResolutionScope : CoroutineContext.Element {
-    override val key: CoroutineContext.Key<*>
-        get() = Key
+interface INodeResolutionScope {
 
+    /**
+     * Implementers should accept instances of NodeReference (a.k.a. SerializedNodeReference) and deserialize them
+     * into supported references if possible. This removes the need to register an INodeReferenceSerializer and also
+     * postpones the deserialization until it's actually necessary.
+     *
+     * For backwards compatability reasons this method should not be called directly, but INodeReference.resolveIn
+     * should be used, which tries to deserialize the reference first, in case an implementation of this method doesn't
+     * do the deserialization itself yet.
+     */
+    @Deprecated("Don't call this method directly. Use INodeReference.resolveIn instead.")
     fun resolveNode(ref: INodeReference): INode?
 
-    companion object Key : CoroutineContext.Key<INodeResolutionScope>
+    /**
+     * All node references inside the body are resolved against this scope. Compared to runWithAlso, the existing scopes
+     * in the current context are not used, meaning they are replaced.
+     */
+    fun <T> runWithOnly(body: () -> T): T = contextScope.computeWith(this, body)
 
-    override fun plus(context: CoroutineContext): CoroutineContext {
-        // coroutines are not compiled with -Xjvm-default=all-compatibility
-        // to not break existing IAreas implemented in Java, this override is necessary
-        return super.plus(context)
-    }
+    /**
+     * Does the same as runWithOnly, but with support for suspendable functions.
+     */
+    suspend fun <T> runWithOnlyInCoroutine(body: suspend () -> T): T = contextScope.runInCoroutine(this, body)
 
-    override fun <R> fold(initial: R, operation: (R, CoroutineContext.Element) -> R): R {
-        // coroutines are not compiled with -Xjvm-default=all-compatibility
-        // to not break existing IAreas implemented in Java, this override is necessary
-        return super.fold(initial, operation)
-    }
+    /**
+     * All node references inside the body are resolved against this scope and if that fails against any other scope in
+     * the current context.
+     */
+    fun <T> runWithAlso(body: () -> T): T = runWithAlso(this, body)
```

Comment:
Same as above. Adding one more word what is "also" included here would make this more readable.